File: step08viewbinding/src/main/java/com/example/step08viewbinding/MainActivity.java
package com.example.step08viewbinding;

import android.os.Bundle;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import com.example.step08viewbinding.databinding.ActivityMainBinding;
/*
view binding 사용하는 방법

1. build.gradle 파일에 설정(아래의 문자열 추가)

    buildFeatures {
        viewBinding = true
    }
2. build.gradle 파일의 우상단의 sync now를 눌러 실제 반영되도록 한다.
3. layout xml 문서의 이름을 따서 자동으로 만들어진 클래스명을 예측한다.
    ex)Activity_xxx.xml 문서의 경우 ActivityxxxBinding이라는 클래스가 만들어진다.
 */

public class MainActivity extends AppCompatActivity implements View.OnClickListener{
    //바인딩 객체의 참조값을 담을 필드
    ActivityMainBinding binding;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setContentView(R.layout.activity_main);
        //activity_main.xml 문서에 전개된 각 객체의 참조값을 얻어오는 방법1
        /*
        EditText editText=findViewById(R.id.editText);
        Button button=findViewById(R.id.button);
        TextView textView=findViewById(R.id.textView);
        */
        //activity_main.xml 문서에 전개된 각 객체의 참조값을 얻어오는 방법2
        //속도가 더 빠르다는 이유로 쓰는 경우가 간혹 있으므로 알아두자.
        /*binding=ActivityMainBinding.inflate(getLayoutInflater());
        EditText editText=binding.editText;
        Button button=binding.button;
        TextView textView=binding.textView;*/

        /*
        EditText의 문자열이 버튼을 누르면 TextView로 이동하도록 프로그래밍한다면 아래와 같은 코딩이 된다.
         */
        //바인딩 객체의 참조값을 얻어와서 필드에 저장
        binding=ActivityMainBinding.inflate(getLayoutInflater());
        //setContentView(R.layout.activity_main); 를 대체함
        setContentView(binding.getRoot());
        //버튼에 리스너 등록하기(바인딩 버전)
        binding.button.setOnClickListener(this);

    }

    @Override
    public void onClick(View view) {
        String msg=binding.editText.getText().toString();

        binding.textView.setText(msg);
    }
}

File: step13sharedpref/src/main/java/com/example/step13sharedpref/MainActivity.kt
package com.example.step13sharedpref

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.PreferenceManager

/*
App에서 문자열을 영구 저장하는 방법(영구 저장이란 앱을 종료하고 다시 시작해도 불러올수 있는 문자열)
1. 파일 입출력을 이용
2. 자체 database를 이용해서 저장
3. sharedPreference를 이용해서 저장(느리지만 간단히 저장하고 불러올수있다)
    내부적으로 xml문서를 만들어서 문자열을 저장하고 불러온다.
    저장된 문자열을 boolean, int, double, String type으로 변환해서 불러올 수 있다.
 */
class MainActivity : AppCompatActivity(), View.OnClickListener {
    /*
    java에서는 field를 선언만 하면 자동으로 null로 초기화되지만
    kotlin에서는 그렇지않다.
    따라서 null이 들어갈 수 있는 field를 따로 지정해서 일시적으로 넣어주어야 한다.
     */
    var editText:EditText?=null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //EditText 객체의 참조값 얻어오기
        editText = findViewById<EditText>(R.id.editText)
        //Button 객체의 참조값 얻어오기
        val saveBtn: Button = findViewById(R.id.saveBtn)
        saveBtn.setOnClickListener(this)

        val readBtn: Button = findViewById(R.id.readBtn)
        //원래 모양은 Object:View.OnClickListener{
        //      override fun onClick(v:View?){
        //     }
        // }의 형태이나, 이를 줄여 쓴 것이다.
        readBtn.setOnClickListener{
            val pref:SharedPreferences = getSharedPreferences("info", Context.MODE_PRIVATE)
            //msg라는 key값으로 저장된 문자열을 읽어오기. 없는 경우를 위해 기본값을 따로 지정해준다.
            val msg=pref.getString("msg","")
            /*여기서 this는 MainActivity 객체를 가리킨다.
            java에서는 익명 클래스 안에서 바깥 클래스의 참조값을 가리키려면
            MainActivity.this 와 같이 클래스명을 따로 명시해야했다

            kotlin에서도 익명 클래스 안쪽에서 바깥 클래스를 가리키려면
            원래는 this@MainClass와 같이 클래스명을 명시해줘야한다.

            단, 간략히 표현한 블럭 안에서는 this만 써도 바깥 클래스의 객체의 참조값이 얻어와진다.
            */
            Toast.makeText(this,msg,Toast.LENGTH_SHORT).show()
        }
    }

    override fun onStart() {
        super.onStart()
        var pref=PreferenceManager.getDefaultSharedPreferences(this)
        //액티비티가 활성화되는 시점에 설정에 저장된 값을 읽어온다.
        val signature=pref.getString("signature","")
        val reply=pref.getString("reply","")
        val sync=pref.getBoolean("sync",false)
        //값이 넘어왔는지 확인
        Toast.makeText(this,signature+reply+sync,Toast.LENGTH_SHORT).show()
    }

    override fun onClick(v: View?) {
        val msg=editText?.text.toString()//null이 가능한 변수인 경우에는 null일수도 있다는 의미에서 ?을 붙여줘야한다.
        //SharedPreference의 참조값 얻어오기
        val pref=getSharedPreferences("info", Context.MODE_PRIVATE)
        //에디터 객체의 참조값 얻어오기
        val editor:SharedPreferences.Editor=pref.edit()
        //에디터 객체를 이용해서 문자열을 key:value 형태로 영구 저장할 수 있다.
        editor.putString("msg", msg)
        editor.commit()

        AlertDialog.Builder(this)
            .setMessage("저장했습니다.")
            .setNeutralButton("확인",null)
            .create()
            .show()
    }
    //옵션 메뉴를 구성하는 메소드
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        //menu/menu_main.xml 문서를 전개해서 메뉴 구성하기
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }
    //옵션 아이템을 선택했을때 호출되는 메소드
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        //선택한 메뉴의 아이디 읽어오기
        val id=item.itemId;
        //만일 설정을 선택한 경우
        if(id == R.id.setting){
            //java라면 SettingsActivity.class였을 것이다.
            val intent= Intent(this, SettingsActivity::class.java)
            //SettingActivity를 시작하겠다는 의도를 전달한다.
            startActivity(intent)
        }
        return super.onOptionsItemSelected(item)
    }
}